### PR TITLE
Add audit history

### PR DESCRIPTION
**Pull request by michielderoos**
**Describe the Pull Request**

Adds audit history tracking and an API to get it!

**Todo**

- [x] Link relevant [trello card](https://trello.com/b/PA2LhlOh/sempo-dev) or [related issue](https://github.com/teamsempo/SempoBlockchain/issues) to the pull request
- [x] Request review from relevant @person or team
- [x] QA your pull request. This includes running the app, login, testing changes, checking [accessibility](https://www.a11yproject.com/checklist/) etc.
- [x] Bump [backend](https://github.com/teamsempo/SempoBlockchain/blob/e3eb0f480e86d5a8ef2c1814127b70ff018671c4/config.py#L7) and/or [frontend](https://github.com/teamsempo/SempoBlockchain/blob/e3eb0f480e86d5a8ef2c1814127b70ff018671c4/app/package.json#L3) versions following Semver
- [ ] Add release notes to [unreleased changelog](https://github.com/teamsempo/SempoBlockchain/blob/master/CHANGELOG.md#unreleased)

### CODE REVIEW

**Comment by tristanhcole** [comment]
[Audit History](https://trello.com/c/J6IxDdFh/887-audit-history)

File: app/server/api/transfer_account_api.py [hunk 1]
```diff
@@ -246,6 +247,20 @@ def put(self):
         }
         return make_response(jsonify(response_object)), 201
 
+class AuditHistoryAPI(MethodView):
+    @requires_auth(allowed_roles={'ADMIN': 'any'}) # Do we want this to be just for superadmins?
```
Comment (tristanhcole) [inline app/server/api/transfer_account_api.py:251]: I think just superadmins for now

File: app/server/api/user_api.py [hunk 2]
```diff
@@ -230,6 +231,20 @@ def post(self, user_id):
             }
             return make_response(jsonify(response_object)), 400
 
+class AuditHistoryAPI(MethodView):
+    @requires_auth(allowed_roles={'ADMIN': 'any'}) # Do we want this to be just for superadmins?
```
Comment (tristanhcole) [inline app/server/api/user_api.py:235]: I think just superadmins for now

File: app/server/models/audit_history.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                    transfer_account_id=transfer_account_id,
+                    user_id=user_id
+                ))
+
+        # We want to add audit history after flush since adding to the database in
+        # before_update is unsafe behavior
+        @event.listens_for(db.session, "after_flush", once=True)
+        def receive_after_flush(session, context):
+            for audit_history in ah:
+                db.session.add(audit_history)
```
Comment (tristanhcole) [inline app/server/models/audit_history.py:54]: This is generally pretty clean, but I'm wondering if we want to make this more generic as a wrapper class? Similar to OneOrgBase, SoftDelete etc.

**Review by tristanhcole — COMMENTED** [review]
Works well! Solid testing and clean code. A few small comments. Maybe also worth considering a Wrapper Class.

Currently not included on user account audit history:
- Custom attributes
- Account types
- Disabled card
- Deleted users/transfer accounts

Do we want to include who deleted the user? This has been a common issue/request from customers

**Review by michielderoos — COMMENTED** [review]
@tristanhcole Refactored it and solved everything! Custom Attributes, Disabled Cards, Account Types, and Deleted Users are all tracked now. There's special logic for deleted users as well-- where history for deleted users is deleted but the record of the deletion itself is kept.

File: app/server/utils/user.py [hunk 4]
```diff
@@ -159,7 +160,6 @@ def update_transfer_account_user(user,
 
     # remove all roles before updating
     user.remove_all_held_roles()
-    flag_modified(user, '_held_roles')
```
Comment (michielderoos) [inline app/server/utils/user.py:162]: flag_modified breaks `load_history().has_changes()`, so I had to cut it out of `user_roles` logic. Everything still works as normal though!

**Comment by tristanhcole** [comment]
When I try to add a card to an existing user, I'm getting this error:

```
Traceback (most recent call last):
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/sentry_sdk/integrations/flask.py", line 90, in sentry_patched_wsgi_app
    environ, start_response
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/sentry_sdk/integrations/wsgi.py", line 129, in __call__
    reraise(*_capture_exception(hub))
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/sentry_sdk/_compat.py", line 55, in reraise
    raise value
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/sentry_sdk/integrations/wsgi.py", line 125, in __call__
    _sentry_start_response, start_response, transaction
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/sentry_sdk/integrations/flask.py", line 89, in <lambda>
    return SentryWsgiMiddleware(lambda *a, **kw: old_app(self, *a, **kw))(
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/flask/app.py", line 2464, in __call__
    return self.wsgi_app(environ, start_response)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/flask/app.py", line 2450, in wsgi_app
    response = self.handle_exception(e)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/flask/app.py", line 1867, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/flask/app.py", line 1953, in full_dispatch_request
    return self.finalize_request(rv)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/flask/app.py", line 1970, in finalize_request
    response = self.process_response(response)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/flask/app.py", line 2267, in process_response
    response = handler(response)
  File "/Users/tristancole/PycharmProjects/SempoBlockchain/app/server/__init__.py", line 167, in after_request
    prepare_transactions_async_job()
  File "/Users/tristancole/PycharmProjects/SempoBlockchain/app/server/utils/executor.py", line 143, in prepare_transactions_async_job
    add_after_request_executor_job(bulk_process_transactions, args=[g.pending_transactions])
  File "/Users/tristancole/PycharmProjects/SempoBlockchain/app/server/utils/executor.py", line 117, in add_after_request_executor_job
    fn.submit(*args, **kwargs)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/flask_executor/executor.py", line 47, in submit
    future = self.executor.submit(self.fn, *args, **kwargs)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/flask_executor/executor.py", line 164, in submit
    future.add_done_callback(callback)
  File "/Users/tristancole/.pyenv/versions/3.7.2/lib/python3.7/concurrent/futures/_base.py", line 403, in add_done_callback
    fn(self)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/flask_executor/executor.py", line 35, in propagate_exceptions_callback
    raise exc
  File "/Users/tristancole/.pyenv/versions/3.7.2/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/flask_executor/executor.py", line 27, in wrapper
    return fn(*args, **kwargs)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/flask/ctx.py", line 158, in wrapper
    return f(*args, **kwargs)
  File "/Users/tristancole/PycharmProjects/SempoBlockchain/app/server/utils/executor.py", line 55, in wrapper
    g.active_organisation = db.session.query(Organisation).filter(Organisation.id == g.active_organisation.id).first()
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/sqlalchemy/orm/attributes.py", line 287, in __get__
    return self.impl.get(instance_state(instance), dict_)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/sqlalchemy/orm/attributes.py", line 718, in get
    value = state._load_expired(state, passive)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/sqlalchemy/orm/state.py", line 652, in _load_expired
    self.manager.deferred_scalar_loader(self, toload)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env10/lib/python3.7/site-packages/sqlalchemy/orm/loading.py", line 946, in load_scalar_attributes
    "attribute refresh operation cannot proceed" % (state_str(state))
sqlalchemy.orm.exc.DetachedInstanceError: Instance <Organisation at 0x11afa1240> is not bound to a Session; attribute refresh operation cannot proceed (Background on this error at: http://sqlalche.me/e/13/bhk3)

```

**Review by tristanhcole — COMMENTED** [review]
`Business Category` custom attribute doesn't seem to be saving in audit history.

Otherwise, tested and all other features seem to be working well!

**Review by tristanhcole — APPROVED** [review]
Tested, LTGM once https://github.com/teamsempo/SempoBlockchain/pull/753#pullrequestreview-822249161 resolved

**Pull request merged**